File: scripts/crosscheck_survey.R
# Independent re-estimation of Part 1 and Part 3 in R's `survey` package.
#
# WHY THIS FILE EXISTS
# --------------------
# It is the other implementation. Everything it computes is already computed in
# `src/descriptive.py` and `src/models.py`; the point is that it was written
# from the published definitions by a different codebase, so when the two agree
# the agreement is evidence and not a tautology. Nothing in here may be adjusted
# to make a number match. If a number does not match, that IS the output.
#
# WHAT WOULD SILENTLY GO WRONG IF THIS FILE WERE SLOPPY
# -----------------------------------------------------
# Three things, each of which produces a plausible-looking wrong answer rather
# than an error:
#
#   nest=TRUE. NHANES numbers PSUs 1,2 INSIDE each stratum. Without nest=TRUE,
#   survey reads every "PSU 1" in the file as the same cluster, pools people
#   from unrelated strata into one ultimate cluster, and returns a standard
#   error that is wrong in an unsignposted direction. The design object is built
#   without complaint either way.
#
#   The standard population is read from part1_stdpop.csv, which Python writes
#   out of STD_2000. It is not typed in here. Two copies of eleven decimal
#   constants in two languages will eventually disagree, and the symptom would
#   be a variance discrepancy in a run whose real fault is a transcription.
#
#   The analytic frames are read, never rebuilt. If this file re-derived the
#   cohort, a disagreement could mean "different formula" or "different people",
#   and telling those apart afterwards is close to impossible.
#
# LONELY PSUs
# -----------
# Run under both survey.lonely.psu="adjust" (the analogue of the collapse rule
# `_linearised_variance` documents: centre the lone unit on the grand mean) and
# "average" (substitute the average within-stratum variance). Both are emitted.
# Where they are identical, the sample has no singleton strata and the project's
# collapse rule is inert for that estimate -- which is worth stating explicitly,
# because "our rule differs from the default" only matters when it fires.
#
#   Rscript scripts/crosscheck_survey.R <exchange_dir> [part1|part3|both]

suppressPackageStartupMessages({
  library(survey)
  library(survival)
})

args <- commandArgs(trailingOnly = TRUE)
if (length(args) < 1L) stop("usage: crosscheck_survey.R <exchange_dir> [part]")
exchange <- args[1L]
part <- if (length(args) >= 2L) args[2L] else "both"

LONELY <- c("adjust", "average")

cat(sprintf("survey %s | survival %s | %s\n",
            packageVersion("survey"), packageVersion("survival"),
            R.version.string))


# ------------------------------------------------------------------ part 1 ---

part1_one <- function(d, std, lonely) {
  options(survey.lonely.psu = lonely)

  # Levels come from the standard-population file, so the contrast vector and
  # the domain means are in the same order by construction rather than by luck.
  d$age_group <- droplevels(factor(d$age_group, levels = std$age_group))
  des <- svydesign(ids = ~psu, strata = ~strata, weights = ~weight,
                   data = d, nest = TRUE)

  # covmat=TRUE keeps the covariance BETWEEN age bands. Dropping it and adding
  # per-band variances is the mistake the Python docstring warns about: one
  # person's PSU contributes to several bands, and treating the bands as
  # independent throws that dependence away silently.
  a <- svyby(~prev_cvd, ~age_group, des, svymean, covmat = TRUE)

  w <- std$std_weight[match(levels(d$age_group), std$age_group)]
  w <- w / sum(w)          # renormalise over the bands present -- Master List
  names(w) <- names(coef(a))
  ctr <- svycontrast(a, list(p_std = w))

  crude <- svymean(~prev_cvd, des)

  # Second, independent R route to the same estimand. svystandardize rescales
  # the weights and post-stratifies; svycontrast linearises a fixed combination
  # of domain means. They should land in the same place, and reporting the gap
  # is how a reader knows the Python number is not being compared against one
  # arbitrary choice among several R answers.
  pop <- std$std_weight[match(levels(d$age_group), std$age_group)]
  sdes <- svystandardize(des, by = ~age_group, over = ~1, population = pop)
  sm <- svymean(~prev_cvd, sdes)

  data.frame(
    lonely_psu   = lonely,
    n            = nrow(d),
    n_psu        = nrow(unique(d[, c("strata", "psu")])),
    n_strata     = length(unique(d$strata)),
    design_df    = degf(des),
    p_std        = as.numeric(coef(ctr)),
    se_std       = as.numeric(SE(ctr)),
    p_std_svystd = as.numeric(coef(sm)),
    se_std_svystd = as.numeric(SE(sm)),
    p_crude      = as.numeric(coef(crude)),
    se_crude     = as.numeric(SE(crude)),
    stringsAsFactors = FALSE
  )
}

run_part1 <- function() {
  d <- read.csv(file.path(exchange, "part1_input.csv"), stringsAsFactors = FALSE)
  std <- read.csv(file.path(exchange, "part1_stdpop.csv"), stringsAsFactors = FALSE)
  cat(sprintf("part1: %d rows, %d cycles\n", nrow(d), length(unique(d$cycle))))

  out <- do.call(rbind, lapply(LONELY, function(lp) {
    do.call(rbind, lapply(sort(unique(d$cycle)), function(cy) {
      # One design per cycle. The stratum codes do not repeat across cycles, so
      # a pooled design subset to a cycle would give the same answer; building
      # per cycle mirrors what by_cycle() does with design=None and removes the
      # question of whether the domain machinery is what is being compared.
      cbind(cycle = cy, part1_one(d[d$cycle == cy, , drop = FALSE], std, lp))
    }))
  }))
  write.csv(out, file.path(exchange, "part1_r.csv"), row.names = FALSE)
  cat(sprintf("part1: wrote %d rows\n", nrow(out)))
}


# ------------------------------------------------------------------ part 3 ---

part3_terms <- function(m, label, design_df = NA_integer_) {
  b <- coef(m)
  se <- sqrt(diag(vcov(m)))

  # NOT confint(m). survey's confint for a svycoxph returns a NORMAL interval,
  # so Part 3's primary interval was built on z = 1.96 while Part 1, the
  # ascertainment series and the Tableau extract all use a t on the design
  # degrees of freedom. Two conventions under one "95% CI" legend, with nothing
  # in the artefact recording which was which -- and design_df was printed to
  # the console rather than written to the file, so the Python loader could not
  # have checked it even if it had wanted to.
  #
  # Pooling eight cycles does buy a lot of df: 123 here, against 8-17 per cycle
  # in Part 1. So the correction is small -- on the exposure, 1.0789-1.1660
  # against 1.0785-1.1664 per 10 mmHg. That is one printed digit in the
  # headline stat card, and being small is not a reason to use a different rule
  # from the rest of the project without saying so.
  crit <- if (is.na(design_df)) qnorm(0.975) else qt(0.975, design_df)
  data.frame(
    fit  = label,
    term = names(b),
    coef = as.numeric(b),
    se   = as.numeric(se),
    lo95 = as.numeric(b - crit * se),
    hi95 = as.numeric(b + crit * se),
    crit = crit,
    # Written, not printed. A multiplier the loader cannot re-derive is a
    # multiplier the loader cannot check.
    design_df = design_df,
    stringsAsFactors = FALSE
  )
}

run_part3 <- function() {
  d <- read.csv(file.path(exchange, "part3_input.csv"), stringsAsFactors = FALSE)
  covs <- readLines(file.path(exchange, "part3_covariates.txt"))
  covs <- covs[nzchar(covs)]
  f <- as.formula(paste("Surv(followup_years, cvd_death) ~",
                        paste(covs, collapse = " + ")))
  cat(sprintf("part3: %d rows, %d events, %d covariates\n",
              nrow(d), sum(d$cvd_death), length(covs)))

  psu_per_stratum <- tapply(d$psu, d$strata, function(x) length(unique(x)))
  cat(sprintf("part3: %d strata, %d singleton strata after listwise deletion\n",
              length(psu_per_stratum), sum(psu_per_stratum < 2)))

  out <- list()
  for (lp in LONELY) {
    options(survey.lonely.psu = lp)
    des <- svydesign(ids = ~psu, strata = ~strata, weights = ~wtmec2yr,
                     data = d, nest = TRUE)
    m <- svycoxph(f, design = des)
    lab <- if (lp == "adjust") "svycoxph" else paste0("svycoxph_", lp)
    ddf <- degf(des)
    if (lp == "adjust") {
      cat(sprintf("part3: design df = %d (PSUs - strata), t = %.5f\n",
                  ddf, qt(0.975, ddf)))
    }
    out[[length(out) + 1L]] <- part3_terms(m, lab, ddf)
  }

  # The same estimator lifelines computes: cluster-robust sandwich, no strata,
  # no n_h/(n_h-1). Fitting it here is what separates "the two implementations
  # disagree" from "the two are estimating different variances on purpose".
  fc <- as.formula(paste("Surv(followup_years, cvd_death) ~",
                         paste(covs, collapse = " + "),
                         "+ cluster(design_cluster)"))
  mc <- coxph(fc, data = d, weights = d$wtmec2yr, robust = TRUE)
  # No design object, so no design df: this one keeps the normal quantile,
  # and the NA in the column is what says so rather than leaving a reader to
  # guess which multiplier produced it.
  out[[length(out) + 1L]] <- part3_terms(mc, "coxph_cluster")

  res <- do.call(rbind, out)
  write.csv(res, file.path(exchange, "part3_r.csv"), row.names = FALSE)
  cat(sprintf("part3: wrote %d rows (%s)\n", nrow(res),
              paste(unique(res$fit), collapse = ", ")))
}


if (part %in% c("part1", "both")) run_part1()
if (part %in% c("part3", "both")) run_part3()

File: scripts/crosscheck_survey.py
"""Cross-validate the hand-written survey estimators against R's `survey` package.

WHAT THIS IS EVIDENCE FOR, AND WHAT IT IS NOT
---------------------------------------------
Every variance estimator in this project is hand-written: the Taylor
linearisation in `src.descriptive` and the cluster-robust Cox in `src.models`.
Both are covered by unit tests, and unit tests written by the author of an
estimator can only check that it does what the author believed it should do.
They cannot catch a shared misreading of the formula, because the test encodes
the same misreading. The only thing that can is an implementation written by
someone else from the same published definitions.

So this script does not test the estimators. It compares them, term by term,
against `survey` -- Lumley's package, which is the reference implementation the
NCHS analytic guidelines and the Stata/SAS survey procedures are checked
against. Agreement to several decimal places is the claim the unit tests cannot
make. Disagreement is a finding, and this script is written to make a
disagreement loud rather than to make the two sides meet: nothing here tunes
the R side toward the Python answer.

WHAT WOULD SILENTLY GO WRONG WITHOUT IT
---------------------------------------
A design-based standard error has no smell. If `_linearised_variance` dropped
the between-age-group covariance, or used n_h rather than n_h/(n_h-1), or
centred PSU totals on the grand mean instead of the stratum mean, the numbers
would still be positive, still be the right order of magnitude, still move in
the right direction across cycles, and still pass a test asserting they are
positive and the right order of magnitude. Every published confidence interval
in Part 1 would be wrong by a factor nobody could see. The same holds for the
Cox robust variance, with one extra trap: lifelines and `survey` compute
DIFFERENT estimators under the same name, and the difference is not a bug in
either. See PART 3 below.

THE EXCHANGE
------------
Python writes the analytic frames it actually fits -- not a re-derivation, the
same rows the shipped code path produces -- into `data/processed/crosscheck/`,
shells out to Rscript, and reads back one CSV per part. The frames are exported
rather than rebuilt in R because rebuilding them in R would put a second cohort
definition in the repository, and then a disagreement could mean either "the
variance formula differs" or "the two sides are looking at different people",
which is the one thing this script must never leave ambiguous.

PART 1: age-standardised prevalence
-----------------------------------
The comparator is `svyby(..., covmat=TRUE)` + `svycontrast`: the same estimand
written the same way, as domain means per age band, their full covariance
matrix, and a fixed linear combination over it. `svystandardize` is run as well
and reported alongside, because it is the route most analysts would take and it
does NOT linearise the same quantity -- it rescales the weights and then takes
an ordinary mean, which treats the within-band weight totals as fixed. Where
the two R routes disagree with each other, that gap is the scale on which
"which R answer are you comparing against" matters, and it belongs in the
output rather than in a footnote.

Both lonely-PSU settings are run. `_linearised_variance` documents a collapse
rule for singleton strata; `survey.lonely.psu="adjust"` centres a lone PSU on
the grand mean, which is the same remedy. The overall by-cycle series has no
singleton strata at all, so if the two settings ever differ for Part 1 then the
export is not the sample that docstring describes and the run is invalid.

PART 3: cause-specific Cox
--------------------------
The coefficients should agree to near machine precision: both sides maximise
the same weighted Efron partial likelihood. If they do not, the disagreement is
in the model frame or in tie handling, not in the variance.

The standard errors are the interesting half, and a difference there is
EXPECTED rather than alarming. lifelines' `cluster_col` sandwich sums squared
cluster-level score residuals with no reference to strata:

    V = sum over clusters c of  u_c u_c'

`survey`'s design-based variance is the stratified ultimate-cluster estimator:

    V = sum over strata h of  n_h/(n_h-1) *
        sum over PSUs i in h of (u_hi - u_h.)(u_hi - u_h.)'

These estimate different things. The first ignores stratification, discarding
the stratification gain; the second removes between-stratum variation and pays
an n_h/(n_h-1) = 2 inflation for it at two PSUs per stratum. Neither is a bug.
But only the second is the design-based variance the NHANES analytic guidelines
ask for, and this project reports its intervals as design-based.

To separate "different estimator" from "different implementation", the R side
also fits plain `coxph` with `cluster()`, which is the SAME estimator lifelines
computes. Python-vs-coxph measures the implementation; coxph-vs-svycoxph
measures the estimator choice. Without that third fit a single number would be
carrying both explanations at once.

    python scripts/crosscheck_survey.py
    python scripts/crosscheck_survey.py --part 1     # one part only
"""

from __future__ import annotations

import argparse
import os
import shutil
import subprocess
import sys
from pathlib import Path

import numpy as np
import pandas as pd

ROOT = Path(__file__).resolve().parent.parent
sys.path.insert(0, str(ROOT))

from src import models  # noqa: E402
from src.descriptive import STD_2000, build_descriptive, by_cycle  # noqa: E402

EXCHANGE = ROOT / "data" / "processed" / "crosscheck"
TABLES = ROOT / "reports" / "tables"
R_SCRIPT = ROOT / "scripts" / "crosscheck_survey.R"
COHORT = ROOT / "data" / "processed" / "cohort_part3.csv.gz"

# R is deliberately NOT assumed to be on PATH -- it is not, on this machine. A
# bare "Rscript" would fail with a message about a missing executable rather
# than about a missing cross-check, which is the kind of failure that gets
# shrugged off and turns into "we ran the crosscheck once, back in August".
RSCRIPT_DEFAULT = r"C:\Program Files\R\R-4.5.2\bin\Rscript.exe"


def rscript_path() -> str:
    """Absolute path to Rscript, overridable for another machine."""
    env = os.environ.get("RSCRIPT")
    if env:
        return env
    if Path(RSCRIPT_DEFAULT).exists():
        return RSCRIPT_DEFAULT
    found = shutil.which("Rscript")
    if found:
        return found
    raise FileNotFoundError(
        f"Rscript not found at {RSCRIPT_DEFAULT}, not on PATH, and $RSCRIPT is "
        "unset. Nothing here can be verified without it -- set RSCRIPT.")


# ---------------------------------------------------------------- part 1 -----

def export_part1() -> pd.DataFrame:
    """Write the Part 1 analytic frame and the Python estimates beside it.

    The standard population is exported too, taken from `STD_2000` itself rather
    than retyped into the R file. Transcribing eleven decimal constants into a
    second language is a silent-drift generator: both sides would still run,
    both would still produce plausible prevalences, and the disagreement would
    be read as a variance problem.
    """
    # `weight` is whichever analysis weight build_descriptive selected; the
    # estimators read that column by name, so exporting the column instead of a
    # hard-coded WTINT2YR/WTMEC2YR keeps the crosscheck pointed at whatever the
    # shipped default currently is rather than at what it was when this was
    # written.
    df = build_descriptive()
    cols = ["cycle", "strata", "psu", "weight", "age_group", "prev_cvd"]
    out = df[cols].copy()
    out["age_group"] = out["age_group"].astype(str)
    out.to_csv(EXCHANGE / "part1_input.csv", index=False)

    std = pd.DataFrame({"age_group": list(STD_2000),
                        "std_weight": list(STD_2000.values())})
    std.to_csv(EXCHANGE / "part1_stdpop.csv", index=False)

    py = by_cycle(df)[["cycle", "n", "n_psu", "p_std", "se_std",
                       "p_crude", "se_crude"]].copy()
    py.to_csv(EXCHANGE / "part1_python.csv", index=False)
    return py


def compare_part1(py: pd.DataFrame) -> pd.DataFrame:
    r = pd.read_csv(EXCHANGE / "part1_r.csv")
    # "adjust" is the setting matching the documented collapse rule. The other
    # setting is carried into the table as evidence that it changed nothing,
    # not as an alternative answer to choose between.
    adj = r[r.lonely_psu == "adjust"].drop(columns=["lonely_psu"])
    adj = adj.rename(columns={c: f"r_{c}" for c in adj.columns if c != "cycle"})
    ave = r[r.lonely_psu == "average"][["cycle", "p_std", "se_std"]]
    ave.columns = ["cycle", "r_p_std_average", "r_se_std_average"]

    m = py.merge(adj, on="cycle", how="outer").merge(ave, on="cycle", how="left")
    for col in ("p_std", "se_std", "p_crude", "se_crude"):
        rcol = f"r_{col}"
        m[f"absdiff_{col}"] = (m[col] - m[rcol]).abs()
        m[f"reldiff_{col}"] = np.where(
            m[rcol] != 0, (m[col] - m[rcol]).abs() / m[rcol].abs(), np.nan)
    return m


# ---------------------------------------------------------------- part 3 -----

def export_part3() -> pd.DataFrame:
    """Write the Part 3 model frame and the lifelines fit beside it.

    The frame is rebuilt here the way `models._fit` builds it and then checked
    against the fitted object's own row count. That assertion carries the whole
    safety of the comparison: if `_fit` ever changes which rows it drops, this
    script would hand R one sample, fit Python on another, and report the
    difference as a variance discrepancy.
    """
    df = pd.read_csv(COHORT)
    d = models.prepare(df, tobin=True)

    # Mirrors fit_aetiologic's covariate assembly rather than restating it, so
    # a change to E2_ADJUSTMENT reaches the crosscheck without a second edit.
    exposure = "systolic_bp"
    covs = [exposure] + [c for c in models.E2_ADJUSTMENT if c != exposure]
    model_cols = covs + ["followup_years", "cvd_death", "wtmec2yr",
                         "design_cluster"]
    fit_df = d[model_cols].dropna()

    cph = models._fit(d, covs, "cvd_death")
    n_fitted = int(cph.weights.shape[0])
    if n_fitted != len(fit_df):
        raise AssertionError(
            f"_fit used {n_fitted} rows, this script exported {len(fit_df)}. "
            "The export mirrors _fit's own dropna and has drifted from it; fix "
            "the mirror before believing any number below.")

    export = d.loc[fit_df.index, model_cols + ["strata", "psu", "cycle"]]
    export.to_csv(EXCHANGE / "part3_input.csv", index=False)
    (EXCHANGE / "part3_covariates.txt").write_text(
        "\n".join(covs) + "\n", encoding="utf-8")

    s = cph.summary
    py = pd.DataFrame({
        "term": s.index,
        "coef": s["coef"].to_numpy(),
        "se": s["se(coef)"].to_numpy(),
        "lo95": s["coef lower 95%"].to_numpy(),
        "hi95": s["coef upper 95%"].to_numpy(),
    })
    py["n"] = n_fitted
    py.to_csv(EXCHANGE / "part3_python.csv", index=False)
    return py


def compare_part3(py: pd.DataFrame) -> pd.DataFrame:
    r = pd.read_csv(EXCHANGE / "part3_r.csv")
    # `crit` comes along so the table records which multiplier each side used.
    # Comparing two intervals that were built with different quantiles and
    # calling the gap a variance difference is the error this column prevents.
    wide = r.pivot(index="term", columns="fit",
                   values=["coef", "se", "lo95", "hi95", "crit"])
    wide.columns = [f"{fit}_{stat}" for stat, fit in wide.columns]
    m = py.set_index("term").join(wide, how="outer").reset_index()

    for fit in ("svycoxph", "coxph_cluster"):
        m[f"absdiff_coef_{fit}"] = (m["coef"] - m[f"{fit}_coef"]).abs()
        m[f"absdiff_se_{fit}"] = (m["se"] - m[f"{fit}_se"]).abs()
        m[f"reldiff_se_{fit}"] = (m["se"] - m[f"{fit}_se"]).abs() / m[f"{fit}_se"]
    m["reldiff_coef_svycoxph"] = np.where(
        m["svycoxph_coef"] != 0,
        (m["coef"] - m["svycoxph_coef"]).abs() / m["svycoxph_coef"].abs(),
        np.nan)
    # The interval is what the report prints, so it is compared as an interval
    # and not left to be inferred from the standard error.
    m["absdiff_lo95_svycoxph"] = (m["lo95"] - m["svycoxph_lo95"]).abs()
    m["absdiff_hi95_svycoxph"] = (m["hi95"] - m["svycoxph_hi95"]).abs()
    return m


# ------------------------------------------------------------------ run ------

def run_r(part: str) -> None:
    exe = rscript_path()
    cmd = [exe, str(R_SCRIPT), str(EXCHANGE), part]
    print(f"  -> {Path(exe).name} {R_SCRIPT.name} {part}", flush=True)
    proc = subprocess.run(cmd, capture_output=True, text=True)
    if proc.stdout.strip():
        print("\n".join("     R| " + ln
                        for ln in proc.stdout.strip().splitlines()))
    if proc.returncode != 0:
        print("\n".join("     R! " + ln
                        for ln in proc.stderr.strip().splitlines()))
        raise RuntimeError(f"Rscript exited {proc.returncode}")
    if proc.stderr.strip():
        # survey emits real warnings here -- lonely PSUs, non-integer weights.
        # They are part of the finding, so they are printed rather than eaten.
        print("\n".join("     R? " + ln
                        for ln in proc.stderr.strip().splitlines()))


def print_part1(m: pd.DataFrame) -> None:
    print("\nPART 1  age-standardised CVD prevalence, per cycle")
    print("  python = src.descriptive.by_cycle | "
          "R = svyby + svycontrast, lonely.psu='adjust'")
    print(f"  {'cycle':<10} {'py p_std':>10} {'R p_std':>10} {'absdiff':>10} "
          f"{'py se':>10} {'R se':>10} {'absdiff':>10} {'rel':>9} {'df':>4}")
    for r in m.itertuples(index=False):
        print(f"  {r.cycle:<10} {r.p_std:>10.6f} {r.r_p_std:>10.6f} "
              f"{r.absdiff_p_std:>10.2e} {r.se_std:>10.6f} {r.r_se_std:>10.6f} "
              f"{r.absdiff_se_std:>10.2e} {r.reldiff_se_std:>9.2e} "
              f"{int(r.r_design_df):>4}")
    print(f"  max |diff|: point {m.absdiff_p_std.max():.3e} | "
          f"se {m.absdiff_se_std.max():.3e} | "
          f"relative se {m.reldiff_se_std.max():.3e}")
    print(f"  crude prevalence max |diff|: point {m.absdiff_p_crude.max():.3e} "
          f"| se {m.absdiff_se_crude.max():.3e}")
    same = np.allclose(m.r_se_std, m.r_se_std_average, rtol=0, atol=1e-12)
    print(f"  lonely.psu adjust == average: {same}   "
          "(expected True: the overall series has no singleton strata)")
    d = (m.r_se_std_svystd - m.r_se_std).abs().max()
    print(f"  svystandardize vs svycontrast, max |se diff|: {d:.3e}")


def print_part3(m: pd.DataFrame) -> None:
    print("\nPART 3  cause-specific Cox for CVD death, "
          "systolic BP + E2 adjustment")
    print("  python = lifelines cluster sandwich | svycoxph = stratified "
          "ultimate cluster | coxph_cluster = lifelines' own estimator, in R")
    print(f"  {'term':<14} {'py coef':>10} {'svy coef':>10} {'absdiff':>9} "
          f"{'py se':>9} {'svy se':>9} {'rel':>8} {'coxph se':>9} {'rel':>8}")
    for r in m.itertuples(index=False):
        print(f"  {r.term:<14} {r.coef:>10.6f} {r.svycoxph_coef:>10.6f} "
              f"{r.absdiff_coef_svycoxph:>9.2e} {r.se:>9.6f} "
              f"{r.svycoxph_se:>9.6f} {r.reldiff_se_svycoxph:>8.1%} "
              f"{r.coxph_cluster_se:>9.6f} {r.reldiff_se_coxph_cluster:>8.1%}")
    print(f"  max |coef diff| vs svycoxph: {m.absdiff_coef_svycoxph.max():.3e}")
    print(f"  se vs svycoxph      : median rel "
          f"{m.reldiff_se_svycoxph.median():.2%}  "
          f"max {m.reldiff_se_svycoxph.max():.2%}")
    print(f"  se vs coxph+cluster : median rel "
          f"{m.reldiff_se_coxph_cluster.median():.2%}  "
          f"max {m.reldiff_se_coxph_cluster.max():.2%}")


def main() -> None:
    ap = argparse.ArgumentParser(description=__doc__.splitlines()[0])
    ap.add_argument("--part", choices=["1", "3", "both"], default="both")
    args = ap.parse_args()

    EXCHANGE.mkdir(parents=True, exist_ok=True)
    TABLES.mkdir(parents=True, exist_ok=True)

    if args.part in ("1", "both"):
        print("PART 1: exporting analytic frame and Python estimates")
        py1 = export_part1()
        run_r("part1")
        m1 = compare_part1(py1)
        m1.to_csv(TABLES / "crosscheck_part1.csv", index=False)
        print_part1(m1)
        print(f"  written: {TABLES / 'crosscheck_part1.csv'}")

    if args.part in ("3", "both"):
        print("\nPART 3: exporting model frame and lifelines fit")
        py3 = export_part3()
        run_r("part3")
        m3 = compare_part3(py3)
        m3.to_csv(TABLES / "crosscheck_part3.csv", index=False)
        print_part3(m3)
        print(f"  written: {TABLES / 'crosscheck_part3.csv'}")


if __name__ == "__main__":
    main()
